### PR TITLE
[extension/oauth2clientauth] Add EndpointParams

### PR DESCRIPTION
**Pull request by tomsanbear**
**Description:** <Describe what has changed.>
Adds support to pass additional token endpoint parameters, in particular support for some oauth providers (in my case Ory Hydra) require the client to request the audience.

**Link to tracking Issue:** N/A

**Testing:** Tested on a custom build of the otel collector, in addition to adding unit test coverage.

**Documentation:** Added description of the endpoint param config in addition to an example in the readme.

### CODE REVIEW

**Comment by pavankrish123** [comment]
@jpkrohling I can take a review for this? Please assign this to me.

**Comment by pavankrish123** [comment]
Thanks @tomsanbear. Can you please modify this test case as well https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/oauth2clientauthextension/extension_test.go#L48  in addition to the configuration test you modified. This test is a proper e2e use case. 

Everything else LGTM

File: extension/oauth2clientauthextension/extension.go [hunk 1]
```diff
@@ -76,6 +76,7 @@ func newClientCredentialsExtension(cfg *Config, logger *zap.Logger) (*ClientCred
 			ClientSecret: cfg.ClientSecret,
 			TokenURL:     cfg.TokenURL,
 			Scopes:       cfg.Scopes,
+			EndpointParams: cfg.EndpointParams,
```
Comment (pavankrish123) [inline extension/oauth2clientauthextension/extension.go:79]: Thanks @tomsanbear. Can you please modify this test case as well https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/oauth2clientauthextension/extension_test.go#L48 in addition to the configuration test you modified. This test is a proper e2e use case.

Everything else LGTM

**Review by pavankrish123 — CHANGES_REQUESTED** [review]

**Comment by jpkrohling** [comment]
> @jpkrohling I can take a review for this? Please assign this to me.

It's yours!

File: extension/oauth2clientauthextension/extension.go (hunk 1 above)
Comment (tomsanbear) [reply extension/oauth2clientauthextension/extension.go:79]: Hey @pavankrish123 , thanks for the quick review, i've pushed the test up to the branch. Let me know if there is anything else you suggest to add :+1:

File: extension/oauth2clientauthextension/extension_test.go [hunk 2]
```diff
@@ -47,11 +47,12 @@ func TestOAuthClientSettings(t *testing.T) {
...
 			name: "all_valid_settings",
 			settings: &Config{
-				ClientID:     "testclientid",
-				ClientSecret: "testsecret",
-				TokenURL:     "https://example.com/v1/token",
-				Scopes:       []string{"resource.read"},
-				Timeout:      2,
+				ClientID:       "testclientid",
+				ClientSecret:   "testsecret",
+				EndpointParams: url.Values{"audience": []string{"someaudience"}},
```
Comment (pavankrish123) [inline extension/oauth2clientauthextension/extension_test.go:52]: Can you please add the equivalent assert to test the `EndpointParams` to complete  the test - https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/549435ea23e9af63abe56945a86a0c34a1c5bba7/extension/oauth2clientauthextension/extension_test.go#L135

File: extension/oauth2clientauthextension/extension.go (hunk 1 above)
Comment (pavankrish123) [reply extension/oauth2clientauthextension/extension.go:79]: Thank you. I have added a minor suggestion on the test. Please add the test and everything else LGTM

File: extension/oauth2clientauthextension/extension_test.go (hunk 2 above)
Comment (pavankrish123) [inline extension/oauth2clientauthextension/extension_test.go:52]: Can you please add the equivalent assert to test the EndpointParams to complete the test - https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/549435ea23e9af63abe56945a86a0c34a1c5bba7/extension/oauth2clientauthextension/extension_test.go#L135

**Review by pavankrish123 — CHANGES_REQUESTED** [review]

File: extension/oauth2clientauthextension/extension_test.go (hunk 2 above)
Comment (tomsanbear) [reply extension/oauth2clientauthextension/extension_test.go:52]: I have update the test case with the assertion, thanks!

**Comment by pavankrish123** [comment]
Thank you! @jpkrohling need help in kicking off the rest of the builds and on merge.

**Review by pavankrish123 — APPROVED** [review]

File: extension/oauth2clientauthextension/README.md [hunk 3]
```diff
@@ -58,6 +60,7 @@ Following are the configuration fields
 - [**token_url**](https://datatracker.ietf.org/doc/html/rfc6749#section-3.2) - The resource server's token endpoint URLs.
 - [**client_id**](https://datatracker.ietf.org/doc/html/rfc6749#section-2.2) - The client identifier issued to the client.
 - [**client_secret**](https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1) - The secret string associated with above identifier.
+- [**enndoint_params**](https://github.com/golang/oauth2/blob/master/clientcredentials/clientcredentials.go#L44) - Additional parameters that are sent to the token endpoint.
```
Comment (jpkrohling) [inline extension/oauth2clientauthextension/README.md:63]:
```suggestion
- [**endpoint_params**](https://github.com/golang/oauth2/blob/master/clientcredentials/clientcredentials.go#L44) - Additional parameters that are sent to the token endpoint.
```

**Review by jpkrohling — APPROVED** [review]

**Comment by tomsanbear** [comment]
Anything we can do to get this merged in? CI fails don't seem related

**Comment by bogdandrutu** [comment]
would be good to rebase, and ensure test pass.

**Comment by bputt-e** [comment]
@tomsanbear would you be able to rebase?

**Comment by tomsanbear** [comment]
Apologies from the spam as a result of the rebase, PR should be up to date with latest main now.

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
Once you add a changelog entry, the CI should pass and I'll get this merged. Thanks for your patience!

**Comment by jpkrohling** [comment]
Sorry, just noticed that the linting is also failing: 

```
golangci-lint run --allow-parallel-runners
extension.go:75: File is not `gofmt`-ed with `-s` (gofmt)
			ClientID:     cfg.ClientID,
			ClientSecret: cfg.ClientSecret,
			TokenURL:     cfg.TokenURL,
			Scopes:       cfg.Scopes,
make[2]: *** [../../Makefile.Common:96: lint] Error 1
```

**Comment by jpkrohling** [comment]
@tomsanbear, sorry to ask this, but the PR seems to have been opened from your `main` branch: could you please fix the conflict? I'll then merge this on green.

**Comment by tomsanbear** [comment]
> @tomsanbear, sorry to ask this, but the PR seems to have been opened from your `main` branch: could you please fix the conflict? I'll then merge this on green.

@jpkrohling updated!

**Pull request merged**